cli/bench/node/path/op_node_path_posix_dirname.js

```diff
@@ -0,0 +1,24 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+import { posix } from "node:path";
+
+const queueMicrotask = globalThis.queueMicrotask || process.nextTick;
+let [total, count] = typeof Deno !== "undefined"
+  ? Deno.args
+  : [process.argv[2], process.argv[3]];
+
+total = total ? parseInt(total, 0) : 50;
+count = count ? parseInt(count, 10) : 10000000;
+
+function bench(fun) {
+  const start = Date.now();
+  for (let i = 0; i < count; i++) fun();
+  const elapsed = Date.now() - start;
+  const rate = Math.floor(count / (elapsed / 1000));
+  console.log(`time ${elapsed} ms rate ${rate}`);
+  if (--total) queueMicrotask(() => bench(fun));
+}
+
+bench(() => {
+  posix.dirname("/a/b");
```

Comment:
@littledivy let me know if this is correct. I saw that other benchmarks did not pass a list of inputs and just looped over the same input so I also didn't add variation.

Reply:
so... this looks pretty bad

```bash
# deno run cli/bench/node/path/op_node_path_posix_dirname.js 
time 134 ms rate 74626865
time 133 ms rate 75187969
time 109 ms rate 91743119
time 111 ms rate 90090090
time 109 ms rate 91743119
time 112 ms rate 89285714
time 111 ms rate 90090090
^C
# ./target/release/deno run cli/bench/node/path/op_node_path_posix_dirname.js
time 2211 ms rate 4522840
time 2245 ms rate 4454342
time 2271 ms rate 4403346
time 2272 ms rate 4401408
time 2280 ms rate 4385964
time 2280 ms rate 4385964
time 2283 ms rate 4380201
```

Reply:
This is honestly not a big surprise. For the most part, for JS this would be just some index searches from a string and then a string slice (pointer plus a few offsets).

For Rust, this is a copy of the string, then some index searches, followed by a string slice, and then finally a new copy.

Reply:
Calling the op with no inputs and just returning `Ok(())` added ~340 ms alone.

Reply:
@20k-ultra could you elaborate on this? Where did you call the op exactly? ~340ms is a horrendous number, last time I checked a "no-op" op shouldn't take more than 1ms.

Reply:
sorry for the confusion @bartlomieju, The ~340 ms is from the benchmark which calls the function 10 million times. It wasn't 340 ms for 1 call of the op.

I made the op function like this just to experiment

```
#[op]
pub fn op_node_path_posix_dirname(path: Value) -> Result<(), AnyError> {
  Ok(())
}
```

results in the following when executing the [benchmark](https://github.com/denoland/deno/pull/17993/files#diff-78e39aa9d2d55ee7df70a488feaf0d59ee27038dfcefc06fedf5a814d075eb03) I added in this PR:

```bash
# ./target/release/deno run cli/bench/node/path/op_node_path_posix_dirname.js
time 704 ms rate 14204545
time 717 ms rate 13947001
time 727 ms rate 13755158
time 728 ms rate 13736263

# deno run cli/bench/node/path/op_node_path_posix_dirname.js
time 130 ms rate 76923076
time 127 ms rate 78740157
time 106 ms rate 94339622
time 106 ms rate 94339622
```

The point I wanted to confirm was that I couldn't beat JS runtime. The deserialization serde is doing to convert the path into a Value might be adding some time too.

Reply:
I suggest to rebase against `main` and pass `path: &str` - with https://github.com/denoland/deno/pull/17996 landed it should be **a lot** faster.

Reply:
switching to &str and rebasing main to get https://github.com/denoland/deno/pull/17996 does improve the time from ~2200 ms mentioned above to ~1500ms!

Since I know now that I have to catch the exception in JS userland to throw `ERR_INVALID_ARG_TYPE` to get the correct code for node compatibility I don't need Value to do the checks.

1500 ms is still a lot so I'll do some more digging.

Reply:
https://github.com/denoland/deno/pull/18051 improved time by about ~60ms in the benchmark.